### PR TITLE
Fix Gemma4 bidirectional attention + add gemma-4-12B (gemma4_unified)

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Two related changes for the Gemma 4 family:

### 1. Fix bidirectional vision-block attention (existing gemma4 models)
The existing gemma4 (32B) and gemma4 MoE multimodal models always built a
purely **causal** decoder, ignoring `use_bidirectional_attention="vision"`.
HuggingFace gemma4 makes contiguous vision-token spans attend
**bidirectionally** (via a blockwise overlay OR-ed onto the causal/sliding
mask, applied to both full- and sliding-attention layers). This adds:

- `use_bidirectional_attention` to `Gemma4Config` + extraction.
- `block_sequence_ids` plumbed embedding → decoder; `create_attention_bias`
  gains a `block_sequence_ids` blockwise-overlay path (forces the float-bias,
  non-GQA attention path with `is_causal=0`).

### 2. Add gemma-4-12B (`gemma4_unified`)
The full encoder-free multimodal `Gemma4UnifiedForConditionalGeneration`:

- **Text backbone**: 48 layers, hidden 3840, dual head_dim (local 256 /
  global 512), `attention_k_eq_v` with a single global KV head, dual RoPE,
  final-logit softcapping. Reuses `Gemma4CausalLMModel`.
- **Vision**: encoder-free embedder — raw patches (6912) → LN → Dense → LN
  → factorized 2D posemb → LN → scale-free RMSNorm → Linear(→3840).
- **Audio**: encoder-free embedder — raw frames (640) → scale-free RMSNorm
  → Linear(→3840).
- Both strip padding inside the ONNX graph (Compress).
- `Gemma4UnifiedModel` + `Gemma4UnifiedTask` build a 3/4-model package
  (decoder, vision_encoder, embedding, optional audio_encoder).

## Testing

Fast suite: 2783 passed. L1 graph-build tests for text and 4-model
multimodal packages.

Real `google/gemma-4-12B` checkpoint, float32 on CUDA (H200):
- **Text prefill** parity vs HF: max_abs_diff ~0.01.
- **Bidirectional mask** parity test.
- **Full multimodal prefill** parity vs HF: vision cos_sim=1.000000,
  decoder last-token cos_sim=1.000000, max_abs_diff=1.10, argmax match,
  no NaN. (Bidirectional reference requires passing `mm_token_type_ids`
  to HF.)

### CODE REVIEW

File: src/mobius/models/gemma4.py [hunk 1]
```diff
@@ -2001,8 +2105,19 @@ def forward(
...
 
+        outputs: dict[str, ir.Value] = {"inputs_embeds": hidden}
+
+        # Vision-block bidirectional attention overlay ids for the decoder.
+        if self._emit_block_sequence_ids:
+            outputs["block_sequence_ids"] = _compute_block_sequence_ids(
+                op,
+                input_ids,
+                image_token_id=self.image_token_id,
+                audio_token_id=self.audio_token_id,
```
Comment (jambayk) [inline src/mobius/models/gemma4.py]: _compute_block_sequence_ids also creates block attention for audio tokens if audio_token_id is not None. But none of the gemma4 models uses block attention for audio tokens and the rest of the code directly checks for "vision". so `audio_token_id` probably should not be provided? for gemma-4-12b it has both audio and vision support and this would not be None.

Comment (justinchuby) [reply src/mobius/models/gemma4.py]: Good catch — you're right. I confirmed against HF: `create_mm_token_type_ids` assigns 0=text, 1=image, 2=video, 3=audio, and `get_block_sequence_ids_for_mask` builds `is_vision = (mm_token_type_ids == 1) | (mm_token_type_ids == 2)`, so audio (type 3) is deliberately excluded from block bidirectional attention.

Fixed in f57b291: dropped the `audio_token_id` branch (and parameter) from `_compute_block_sequence_ids` so only image runs form blocks — audio placeholders now keep plain causal attention, matching HF. gemma4_unified has no video modality, so this reduces to image tokens alone. Updated the two unit tests accordingly (audio runs now expect -1).

File: src/mobius/_configs/_base.py [hunk 2]
```diff
@@ -1267,6 +1267,17 @@ class Gemma4Config(VisionLanguageConfig):
...
+    use_bidirectional_attention: str | None = None
+    """Bidirectional attention mode for the text decoder.
+
+    Mirrors HF ``Gemma4TextConfig.use_bidirectional_attention``:
+    - ``None``: fully causal (smaller Gemma4 models, e.g. E2B).
+    - ``"vision"``: text stays causal, but contiguous vision-token blocks
+      (image/audio placeholders) attend bidirectionally within each block
+      (larger models, e.g. 12B/26B/32B). Implemented via a per-position
+      ``block_sequence_ids`` overlay added onto the causal mask.
+    - ``"all"``: every token attends bidirectionally (no causal mask).
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/_configs/_base.py]: Fixed in fb0ad6b — updated the docstring to state that `"all"` is not implemented (treated as fully causal; only `"vision"` activates the block overlay). Also corrected the `"vision"` description: only image blocks are bidirectional, audio is excluded (it's token-type 3 in HF's mm_token_type_ids).

File: tests/integration_test.py [hunk 3]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
+    seq_len = input_ids.shape[1]
+    position_ids = np.arange(seq_len, dtype=np.int64)[np.newaxis, :]
+
+    with torch.no_grad():
+        hf_out = hf_full(
+            input_ids=torch.from_numpy(input_ids),
+            attention_mask=torch.from_numpy(attention_mask),
+            position_ids=torch.from_numpy(position_ids),
+        )
+    hf_logits = hf_out.logits.numpy()
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py]: Fixed in fb0ad6b — converted the HF reference tensors with `.detach().cpu().numpy()` so the parity check works regardless of CPU/CUDA placement.

File: tests/integration_test.py [hunk 4]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
+            mm_token_type_ids=mm_token_type_ids,
+        )
+        hf_image_features = (
+            hf_full.model.get_image_features(
+                pixel_values, image_position_ids, return_dict=True
+            )
+            .pooler_output.numpy()
+            .astype(np.float32)
+        )
+    hf_logits = hf_out.logits.numpy()
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py]: Fixed in fb0ad6b — converted the HF reference tensors with `.detach().cpu().numpy()` so the parity check works regardless of CPU/CUDA placement.

File: tests/integration_test.py [hunk 5]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
+            attention_mask=torch.from_numpy(attention_mask),
+            pixel_values=pixel_values,
+            image_position_ids=image_position_ids,
+            mm_token_type_ids=mm_token_type_ids,
+        )
+        hf_image_features = (
+            hf_full.model.get_image_features(
+                pixel_values, image_position_ids, return_dict=True
+            )
+            .pooler_output.numpy()
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py]: Fixed in fb0ad6b — converted the HF reference tensors with `.detach().cpu().numpy()` so the parity check works regardless of CPU/CUDA placement.

File: tests/integration_test.py [hunk 6]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
+    # boa/image/audio token ids and use_bidirectional_attention="vision").
+    text_cfg = hf_config.text_config
+    gemma4_config = Gemma4Config.from_transformers(text_cfg, parent_config=hf_config)
+    gemma4_config = dataclasses.replace(gemma4_config, dtype=ir.DataType.FLOAT)
+
+    onnx_module = Gemma4CausalLMModel(gemma4_config)
+    pkg = build_from_module(onnx_module, gemma4_config, task="gemma4-text-generation")
+    assert "model" in pkg
+
+    preprocessed = onnx_module.preprocess_weights(dict(hf_full.state_dict()))
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py:5404]: Checked this against the current HF `Gemma4ForConditionalGeneration` layout: `language_model` is a `Gemma4TextModel` (not a `*ForCausalLM`), so its keys are `model.language_model.layers.*` / `model.language_model.embed_tokens.weight` — the substring replace yields `model.layers.*` / `model.embed_tokens.weight`, which match the ONNX initializer names. The L4 parity test confirms this (last-token cosine 1.0 vs HF). There's no extra `.model.` segment in this version, so weights are not skipped. Agreed it would be more robust to pull the submodule state_dict directly; happy to switch if a future HF layout reintroduces the extra prefix.

File: tests/integration_test.py [hunk 7]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
+
+    if not _model_accessible(model_id):
+        pytest.skip(f"{model_id} not accessible (requires HuggingFace authentication)")
+
+    hf_config = transformers.AutoConfig.from_pretrained(model_id)
+    hf_full = AutoModelForImageTextToText.from_pretrained(
+        model_id,
+        dtype=torch.float32,
+        low_cpu_mem_usage=True,
+    ).eval()
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py:5391]: Acknowledged. These run the f32 reference on CPU; on the H200 box the 12B forward completes (parity cosine 1.0), so it's not timing out in practice, but loading the reference on `_get_test_device()` would be faster and match the other multimodal tests. Tracking as a follow-up improvement rather than blocking this PR.

File: tests/integration_test.py [hunk 8]
```diff
@@ -5344,3 +5344,403 @@ def test_gemma4_e2b_text_prefill_bf16():
...
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/integration_test.py:5496]: Acknowledged. These run the f32 reference on CPU; on the H200 box the 12B forward completes (parity cosine 1.0), so it's not timing out in practice, but loading the reference on `_get_test_device()` would be faster and match the other multimodal tests. Tracking as a follow-up improvement rather than blocking this PR.

File: tests/build_graph_test.py [hunk 9]
```diff
@@ -1249,6 +1249,57 @@ def test_gemma4_moe_graph(self):
         assert "input_ids" in input_names
         assert "logits" in output_names
 
+    def test_gemma4_unified_text_graph(self):
+        """Build the gemma4_unified (gemma-4-12B) text backbone via Gemma4CausalLMModel.
+
+        ``gemma4_unified_text`` reuses Gemma4CausalLMModel.  This exercises the
+        12B-family text architecture: dual head_dim (local 16 / global 32),
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply tests/build_graph_test.py:1256]: Good point — `gemma4_unified_text` currently only has the dedicated graph-build test and misses the L3 synthetic-parity / weight-alignment coverage driven by `_test_configs.py`. Tracking as a follow-up to add a tiny config entry; the architecture is exercised today via the gemma4_unified multimodal graph + the 12B integration parity tests.

File: examples/gemma4_unified_ort_genai.py [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+        dtype: Model dtype (``"f16"`` recommended for CUDA).
+        ep: Execution provider for ``genai_config.json`` session options.
+    """
+    from mobius.integrations.ort_genai import auto_export
+
+    print(
+        f"Building unified multimodal package for {model_id!r} "
+        f"(dtype={dtype}, ep={ep}) — this downloads ~25GB of weights ..."
+    )
+    manifest = auto_export(model_id, output_dir, dtype=dtype, ep=ep)
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply examples/gemma4_unified_ort_genai.py:144]: False positive: `auto_export` is a function re-exported from `mobius.integrations.ort_genai.__init__` (the static analyzer resolves it to the same-named submodule). The example runs end-to-end and produces the package, confirming it's callable.

File: src/mobius/models/gemma4_test.py [hunk 11]
```diff
@@ -176,3 +176,283 @@ def test_vnorm_fp16_no_nan(self):
...
+                mm_embed_dim=3840,
+                patch_size=16,
+                pooling_kernel_size=3,
+                mm_posemb_size=1120,
+                output_proj_dims=3840,
+                rms_norm_eps=1e-6,
+            ),
+        )
+        fields: dict = {}
+        _gemma4_unified_vision(composite, None, "gemma4_unified", fields)
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/gemma4_test.py:430]: False positive: `@register_vision_hook` returns the original function unchanged, so `_gemma4_unified_vision` keeps its real 4-arg signature `(config, parent_config, model_type, fields)`. The analyzer mistakenly resolves the call target to the decorator's inner 1-arg wrapper. These tests pass.

**Comment by titaiwangms** [comment]
## Review synthesis (4-reviewer team: readability, code, critical, deep)

**Verdict: Strong PR, no Critical issues.** The deep reviewer verified the bidirectional block-overlay mask math is bit-for-bit faithful to HF's blockwise overlay (grounded on Gemma3's identical algorithm, since Gemma4 isn't in the installed transformers). Findings below are deduplicated; the deep reviewer's grounded verdict wins on math/spec ties (it downgraded the cache-boundary concern from Major to Minor — safe for single-shot prefill where `past=0`).

### Major
1. **`TypeError` in an integration test** — `tests/integration_test.py:2721` calls `_compute_block_sequence_ids(op, iid, image_token_id=..., audio_token_id=None)`, but the signature is `(op, input_ids, *, image_token_id)`. The stray `audio_token_id` kwarg raises on run. It is `@integration`-gated, so the fast suite (2783 passed) never exercised it. Remove the kwarg.
2. **`auto_export` routes `gemma4_unified` through the wrong Gemma4 processors** (`src/mobius/integrations/ort_genai/auto_export.py`) — emits SigLIP-style `Gemma4ImageTransform` and `Gemma4LogMel`/128-dim audio configs, but the unified model is encoder-free (raw merged patches / 640-dim frames). The example bypasses this, but `auto_export()` still emits a misleading/broken processor contract. Either omit native image/audio processor configs for unified (with an explicit unsupported marker) or generate dedicated unified processor configs.
3. **Overlay disables GQA for the entire text/decode phase** (`src/mobius/models/gemma4.py:~1644`) — `use_block_overlay` becomes true whenever `use_bidirectional_attention=="vision"` and `input_ids` is present, even with zero image tokens. This forces the slow float-bias `is_causal=0` path for text-only prompts and every decode step, losing the fused/cache-friendly path for a 12B model. Avoid enabling the overlay when there are no vision tokens, or split prefill-overlay from decode.
4. **Test docstring contradicts its own assertion** (`tests/build_graph_test.py:~2279`, `test_gemma4_unified_multimodal_graph`) — the docstring says the embedding model emits `block_sequence_ids`, but the test asserts `"block_sequence_ids" not in e_outputs`. Stale remnant of an earlier design; fix the docstring to state the decoder derives the overlay internally.

### Minor
- **Config hooks silently default architecture-defining dims** (`_configs/per_model/_gemma4_unified_{vision,audio}.py`: `mm_embed_dim=3840`, `mm_posemb_size=1120`, `audio_embed_dim=640`). For `gemma4_unified*`, validate required fields and fail fast instead of building a plausible-but-wrong graph if HF field names drift.
- **Audio `rms_norm_eps` ignored** — `_Gemma4UnifiedAudioEmbedderModel` uses the top-level text `config.rms_norm_eps` rather than `audio_config.rms_norm_eps`; map and use the audio value so audio/text eps can differ.
- **`use_bidirectional_attention=="all"` silently degrades to causal** (`_configs/_base.py` + `gemma4.py:~1117`). HF `"all"` makes every token bidirectional; this produces a causal model instead. Raise `NotImplementedError` so a mis-specified config fails loudly.
- **kv_group left-pad assumes non-chunked prefill** (`components/_common.py:~753`) — left-padding cached positions with `-1` is correct for `past=0`, but cross-chunk intra-image bidirectionality would be lost if prefill is ever chunked. Add a comment documenting the assumption.
- **`image_token_id or 0` fallback** (`gemma4.py:~1440`) is a latent footgun — a missing/zero image token id would mark token 0 as a vision block (and `0` also disables the overlay). Consider asserting it is populated and non-zero.
- **Decode-path test is effectively inert** (`components/_common_test.py:872`) — `test_decode_single_query_is_causal` uses q_group `-1`, so the new `kv_group` left-pad branch isn't meaningfully validated. Add a decode case with `q_len < total_len` and a real block id.
- **Type/clarity**: `use_bidirectional_attention: str | None` → `Literal["vision", "all"] | None`; dead `if vc else None` guards (vision embedder is only built when `config.vision` is populated) → `assert vc is not None`.

### Open questions for the author (HF-grounded, from the deep reviewer)
- Confirm `output_proj_dims == mm_embed_dim` for the real `gemma-4-12B` config — the hard assert (`gemma4.py:~1346`) treats `output_proj_dims` as the projection *input* and passes only coincidentally (both 3840). If HF's `output_proj_dims` is the text hidden size and differs, the build aborts.
- Confirm 12B/26B/32B configs use `"vision"` and never `"all"` (see the silent-fallback Minor above).
- Confirm Gemma4's `get_block_sequence_ids_for_mask` truly reduces to images-only — the equivalence proof was grounded on Gemma3's `token_type_ids_mask_function` since Gemma4 wasn't in the installed transformers.

### Nit
Duplicate `_UNIFIED_TYPES` name across the two hook files (rename per-modality); the `bidirectional`/`use_block_overlay` two-variable chain (e.g. `block_overlay_configured`/`block_overlay_active`); `attn_is_causal: int` deserves a `# 0/1 ONNX attribute, not a bool` comment; the per-layer KV-feed construction is duplicated verbatim in two integration tests.

### Praise
`_compute_block_sequence_ids` is exceptionally documented (HF-cited, every op shape-annotated). The fp16 patch_dense overflow handling (`_F32Linear`/`_F32LayerNorm`, engaged only for FLOAT16) is numerically sound and co-located at the source. `TestBlockwiseAttentionBias` runs four precise ORT-backed invariants (multi-block bidirectionality, sliding-window override, padding, decode). The deep reviewer confirmed the overlay math and OR-before-padding-AND ordering match HF exactly.

<sub>Synthesized from a 4-model review team (readability, code, critical, deep reviewers).</sub>

**Comment by justinchuby** [comment]
“Overlay disables GQA for the entire text/decode phase” --> this isn't applicable because the graph is static.
